### DISCUSSION TITLE
Save forms to bitmaps?

### DISCUSSION THREAD

**Discussion by fredvs** [mse-org/mseide-msegui#62] [General]
There is a topic on Lazarus forum about save forms to bitmaps with MSEgui:
https://forum.lazarus.freepascal.org/index.php/topic,64187.msg487549.html#msg487549

This is the comment:

Have a look at MSEide/MSEgui. For now it can't do that either but compared to LCL's customdrawn, MSEgui is much closer to achieving that.
https://github.com/mse-org/mseide-msegui
Unfortunately, at the moment, unlike LCL, MSEgui is not able to save forms to bitmaps, which is something very important for MSEgui. Much more important than it is for LCL.

-----------------------

It seems to me that it was already asked but cannot remember.

Is it possible to save forms to bitmaps?
Sure with a capture-screen and select a position it is possible but it seems they want something else.

**Comment by fredvs** [comment]
@rchastain : In one of your explorations you found an MSEgui feature of TWindow but I can't seem to find it.
You also show msegui+Cairo: https://gitlab.com/rchastain/msegui-studies/-/tree/master/cairo?ref_type=heads

I think that we have the keys to make MSEgui totally universal.
That said, it wouldn't surprise me if Martin had foreseen it and that a hidden option remains to be discovered.

**Comment by fredvs** [comment]
Hello everybody.

Finally I find a way to save the canvas of a form, with all his child-components into a png file (or/and bitmap).
Only the canvas is saved, even if the form is covered by other external forms.

Here the trick (thanks to the wonderful BGRABitmap gem):

----------------------------

```
procedure tmainfo.saveform2png(const Sender: TObject);
var
  btmp: tbgrabitmap;
  r: rectty;
  p: pointty;
begin
 
 btmp := tbgrabitmap.Create(Width, Height);

  r.x  := 0;
  r.y  := 0;
  r.cx := Width;
  r.cy := Height;

  p.x := 0;
  p.y := 0;

  btmp.Canvas.copyarea(getcanvas, r, p);  // The Trick!

  btmp.savetofile('aBitmap.png');

  btmp.Free;

end;
```

--------------------------------

Here zip file demo:

[msegui_form_savetofile.zip](https://github.com/mse-org/mseide-msegui/files/15342962/msegui_form_savetofile.zip)

**Reply by rchastain** [reply]
Hello Fred. Good job.

Here the image has small imperfections, I don't know why.

![aBitmap](https://github.com/mse-org/mseide-msegui/assets/13957204/b5fb6cf0-80bd-4db7-a00b-90a1cb22b982)

And, by the way, I think there's no reason use Canvas methods with MSEgui, since they are provided for compatibility with the LCL.

Have a nice day.

**Reply by fredvs** [reply]
Hello Roland.

Sorry but could you show me the imperfection, I dont see it.

>And, by the way, I think there's no reason use Canvas methods with MSEgui, since they are provided for compatibility with the LCL.

Hum, this one I dont catch, why Canvas is provided for LCL compatibility?
Canvas is used a lot of time in msegui source.

But surely I miss something, please explain.

**Reply by rchastain** [reply]
I mean _MyBGRABitmap.Canvas.Method_. These methods are done to replace easily LCL canvas methods.

**Reply by fredvs** [reply]
Do you see a other way?

**Comment by fredvs** [comment]
Hello Roland.

I try to find the imperfection with the form saved from a screenshot vs from form2file() but, apart the decorations that of course are not rendered with form2file(), I dont see it.   Please show me what I miss. 

(Ha, ok, there is the "B" of "Boolean edit" that is not perfect. For the button "Save form to file", the frame is not the same but that is because form2file() was called after button.onexecute.)

Form from screenshot:
![screenshot_form2file](https://github.com/mse-org/mseide-msegui/assets/3421249/b1f82ba1-ed1c-439e-8f2b-e9ce4ec80c4c)

Form from form2file():
![aBitmap](https://github.com/mse-org/mseide-msegui/assets/3421249/19148194-d76d-4425-917a-e89b0731ba9b)

**Comment by fredvs** [comment]
Here zoom 4x of "Boolean edit" text.
Indeed they are not the same, it seems that there is something not perfect with the rendering of the mask.

4x zoom from screenshot:
![bool_screenshot](https://github.com/mse-org/mseide-msegui/assets/3421249/20a9009d-2dcb-43cb-bc5b-09ba2a774224)

4x zoom from form2file():
![bool_form2file](https://github.com/mse-org/mseide-msegui/assets/3421249/9405ccdf-b3e3-443b-a874-201cfdaa95d1)

Strange....

**Comment by rchastain** [comment]
> Strange....

The solution is to add a black background:

```pascal
var
  btmp, btmp2: tbgrabitmap;
  // ...
begin
  btmp := tbgrabitmap.Create(Width, Height);
  btmp2 := tbgrabitmap.Create(Width, Height, BGRABlack);
  
  // ...

  btmp.Canvas.copyarea(getcanvas, r, p);
  btmp2.PutImage(0, 0, btmp, dmDrawWithTransparency);
  
  btmp2.savetofile('aBitmap.png');
```

**Reply by fredvs** [reply]
Magnifique!

**Reply by fredvs** [reply]
Thanks Roland to make things perfect.
Here the project updated.
[msegui_form_savetofile.zip](https://github.com/mse-org/mseide-msegui/files/15370766/msegui_form_savetofile.zip)
